rangeserver/src/cache/memtabledb.rs

```diff
@@ -0,0 +1,643 @@
+    }
+}
+
+impl Ord for MemTableEntry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let key_comp = self.key.cmp(&other.key);
+        match key_comp {
+            Ordering::Less | Ordering::Greater => key_comp,
+            Ordering::Equal => other.epoch.cmp(&self.epoch),
+        }
+    }
+}
+
+impl PartialEq for MemTableEntry {
+    fn eq(&self, other: &Self) -> bool {
+        self.key == other.key && self.epoch == other.epoch
+    }
+}
+
+#[derive(Debug)]
+pub struct MemTable {
+    id: usize,
+    size_in_bytes: u64,
+    status: MemTableStatus,
+    latest_epoch: u64,
+    oldest_epoch: u64,
+    entries: OrderedSkipList<MemTableEntry>,
+}
+
+#[derive(Debug)]
+pub struct MemTableDB {
+    rc_options: RCOptions,
+    memtables: Vec<MemTable>,
+    mutable_id: usize,
+    latest_epoch: u64,
+    gc_callback: Option<GCCallback>,
+}
+
+impl MemTableDB {
+    pub async fn new(rc_options: Option<&RCOptions>) -> Self {
+
+        let mut opts = RCOptions{
+            path: "",
+            num_write_buffers: 2,
+            write_buffer_size: 2<<29,
+         };
+
+        match rc_options {
+            Some(opt) => {opts = *opt},
+            None => {},
+        }
+
+        let mut mt_db = MemTableDB {
+            rc_options: opts.clone(),
+            memtables: Vec::<MemTable>::new(), 
+            mutable_id: 0,
+            latest_epoch: 0,
+            gc_callback: None,
+        };
+
+        for i in 0..opts.num_write_buffers {
+            let mt = MemTable { 
+                id: i,
+                size_in_bytes: 0,
+                status: MemTableStatus::FLUSHED,
+                latest_epoch: 0,
+                oldest_epoch: 0,
+                // TODO: init with capacity
+                entries: OrderedSkipList::<MemTableEntry>::new(),};
+            mt_db.memtables.push(mt);
+        }
+        // change the status of first memtable as mutable
+        mt_db.memtables[mt_db.mutable_id].status = MemTableStatus::MUTABLE;
+        mt_db
+    }
+
+    async fn upsert_or_delete(
+        &mut self,
+        key: bytes::Bytes,
+        epoch: u64,
+        deleted: bool,
+        val: Option<bytes::Bytes>) -> Result<(), Error> {
+
+            // sanity checks
+            assert!(self.mutable_id >= 0 && self.mutable_id < self.rc_options.num_write_buffers, 
+                "mutable_id is out of range ({})", self.mutable_id);
+            
+            let mut val_match = Bytes::new();
+            let mut val_len = 0; 
+            match val{
+                Some(v) => {
+                    val_len = v.len();
+                    val_match = v;
+                },
+                None => assert!(deleted == true, "upsert api called without value argument"),
+            }
+
+            let entry_size: u64 = key.len() as u64 + (val_len as u64) + (mem::size_of::<u64>() as u64) + (mem::size_of::<bool>() as u64);
+            let entry = MemTableEntry { key: key, val: val_match, epoch: epoch, deleted: deleted };
```

Comment:
Can we use the [field init shorthand](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#using-the-field-init-shorthand)?